LogSoftMax.cu

```diff
@@ -1,173 +1,297 @@
+          float defaultVal) {
+  float threadVal = defaultVal;
+  int offset = threadIdx.x;
+
+  int last = size % (ILP * blockDim.x);
+
+  // Body (unroll by ILP times)
+  for (; offset < size - last;
+       offset += blockDim.x * ILP) {
+  float tmp[ILP];
+
+#pragma unroll
+    for (int j = 0; j < ILP; ++j) {
+      tmp[j] = data[offset + j * blockDim.x];
+    }
+
+#pragma unroll
+    for (int j = 0; j < ILP; ++j) {
+      threadVal = r(threadVal, tmp[j]);
+    }
+  }
 
-__global__ void cunn_LogSoftMax_updateGradInput_kernel(float *gradInput, float *output, float *gradOutput, int nframe, int dim)
-{
-  __shared__ float buffer[LOGSOFTMAX_THREADS];
-  int k = blockIdx.x;
-  float *gradInput_k = gradInput + k*dim;
-  float *output_k = output + k*dim;
-  float *gradOutput_k = gradOutput + k*dim;
-  int tx = threadIdx.x;
+  // Epilogue
+  for (; offset < size; offset += blockDim.x) {
+    threadVal = r(threadVal, data[offset]);
+  }
 
-  int i_end = dim;
-  int i_step = blockDim.x;
+  return threadVal;
+}
 
-  // sum?
-  buffer[tx] = 0;
-  for (int i=tx; i<i_end; i+=i_step)
-    buffer[tx] += gradOutput_k[i];
+template <int ILP>
+__global__ void
+cunn_LogSoftMax_updateOutput_kernel(float *output, float *input, int classes) {
+  extern __shared__ float buffer[];
+  input += blockIdx.x * classes;
+  output += blockIdx.x * classes;
+
+  float threadMax =
+    ilpReduce<MaxFloat, ILP>(input, classes, MaxFloat(), -FLT_MAX);
+  float max_k =
+    blockReduce<MaxFloat>(buffer, threadMax, MaxFloat(), -FLT_MAX);
+
+  float threadExp =
+    ilpReduce<SumExpFloat, ILP>(input, classes, SumExpFloat(max_k), 0.0f);
+  float logsum_k =
+    blockReduce<SumFloat, LSMFinal>(
+      buffer, threadExp, SumFloat(), 0.0f, LSMFinal(max_k));
+
+  // Output LSM (hand ILP)
+  int offset = threadIdx.x;
+
+  int last = classes % (ILP * blockDim.x);
+  for ( ; offset < classes - last; offset += blockDim.x * ILP) {
+    float tmp[ILP];
+
+#pragma unroll
+    for (int j = 0; j < ILP; ++j) {
+      tmp[j] = input[offset + j * blockDim.x];
+    }
+
+#pragma unroll
+    for (int j = 0; j < ILP; ++j) {
+      output[offset + j * blockDim.x] = tmp[j] - logsum_k;
+    }
+  }
 
-  // reduce
-  for (unsigned int stride = blockDim.x >> 1; stride > 0; stride >>= 1)
-  {
-    __syncthreads();
-    if (tx < stride)
-      buffer[tx] += buffer[tx+stride];
+  for (; offset < classes; offset += blockDim.x) {
+    output[offset] = input[offset] - logsum_k;
   }
+}
 
-  __syncthreads();
+template <int ILP>
+__global__ void
+cunn_LogSoftMax_updateGradInput_kernel(float *gradInput,
+                                       float *output,
+                                       float *gradOutput,
+                                       int classes) {
+  extern __shared__ float buffer[];
+  gradInput += blockIdx.x * classes;
+  output += blockIdx.x * classes;
+  gradOutput += blockIdx.x * classes;
+
+  float threadSum =
+    ilpReduce<SumFloat, 4>(gradOutput, classes, SumFloat(), 0.0f);
```

Comment:
<SumFloat, 4> should be <SumFloat, ILP> I think. Harmless (not a correctness bug) but I think an ILP level of 2 was optimal when I tested it, which is what this is instantiated at. This was probably left over from a previous version.